c2rust-analyze/src/c_void_casts.rs

```diff
@@ -0,0 +1,378 @@
+    /// That is, find all calls to:
+    ///
+    /// * `malloc`
+    /// * `calloc`
+    /// * `realloc`
+    /// * `free`
+    ///
+    /// and insert their casts to and from [`*c_void`].
+    ///
+    /// [`*c_void`]: core::ffi::c_void
+    pub fn insert_all_from_mir(&mut self, mir: &Body<'tcx>, tcx: TyCtxt<'tcx>) {
+        for bb_data in mir.basic_blocks().iter() {
+            if let Some(term) = &bb_data.terminator {
+                let term: &Terminator = term;
+                if let TerminatorKind::Call {
+                    ref func,
+                    ref args,
+                    destination,
+                    target,
+                    ..
+                } = term.kind
+                {
+                    let func_ty = func.ty(&mir.local_decls, tcx);
+
+                    // For [`CVoidCastDirection::From`], we only count
+                    // a cast from `*c_void` to an arbitrary type in the subsequent block,
+                    // searching forward.
+                    let find_first_cast_succ_block = |get_cast| {
+                        let successor_block_id = target.unwrap();
+                        mir.basic_blocks()[successor_block_id]
+                            .statements
+                            .iter()
+                            .find_map(get_cast)
+                    };
+
+                    // For [`CVoidCastDirection::From`], we only count
+                    // a cast to `*c_void` from an arbitrary type in the same block,
+                    // searching backwards.
+                    let find_last_cast_curr_block =
+                        |get_cast| bb_data.statements.iter().rev().find_map(get_cast);
+
+                    for direction in CVoidCastDirection::from_callee(ty_callee(tcx, func_ty))
+                        .iter()
+                        .copied()
+                    {
+                        use CVoidCastDirection::*;
+                        let c_void_ptr = match direction {
+                            From => destination,
+                            To => args[0].place().unwrap(),
+                        };
+                        let c_void_ptr = CVoidPtr::checked(c_void_ptr, &mir.local_decls, tcx);
+                        let get_cast = move |stmt: &Statement<'tcx>| {
+                            c_void_ptr.get_cast_from_stmt(direction, stmt)
+                        };
+                        let cast = match direction {
+                            From => find_first_cast_succ_block(get_cast),
+                            To => find_last_cast_curr_block(get_cast),
+                        };
+                        if let Some(cast) = cast {
+                            self.insert(direction, cast);
+                        }
+                    }
+                }
+            }
+        }
+    }
+
+    pub fn new(mir: &Body<'tcx>, tcx: TyCtxt<'tcx>) -> Self {
+        let mut this = Self::default();
+        this.insert_all_from_mir(mir, tcx);
```

Comment:
why keep `insert_all_from_mir`? that function also wouldn't need to stay public afaict

Reply:
Fixed in 4326435687e15f1e266f52545b78a4daed6166fb.

I had thought `insert_all_from_mir` could be callable outside of `new`, but since `Place`s are local to a func `Body`, that shouldn't happen, so now `CVoidCasts::new` is the only way to create and modify a `CVoidCasts`.

I also renamed `mir` to `body`, as I'm not sure why we just call a `Body` `mir`.